File: stream-analyse/src/main/scala/com/dbapp/analyse/stream/jobs/DBAuditJob.scala
package com.dbapp.analyse.stream.jobs

import java.util.Properties

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.streaming.api.scala.{StreamExecutionEnvironment, _}
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer

case class DBAuditLog (logType: String,
                       startTime: String,
                       sqlLen: String,
                       clientUserName: String,
                       srcHostName: String,
                       dvcAction: String,
                       destMacAddress: String,
                       opType: String,
                       databaseName: String,
                       srcMacAddress: String,
                       destAddress: String,
                       logSessionId: String,
                       alarmFlag: String,
                       responseCode: String,
                       destPort: String,
                       srcUserName: String,
                       relateIp: String,
                       payload: String,
                       severity: String,
                       srcAddress: String,
                       clientPrg: String,
                       dataSet: String,
                       relateUser: String,
                       tenant: String,
                       sqlId: String,
                       costTime: String,
                       databaseType: String,
                       ruleName: String,
                       errorMessage: String,
                       accessId: String,
                       relateUserInfo: String,
                       effectRow: String,
                       srcPort: String,
                       dataSetSize: String,
                       relateUrl: String,
                       deviceAddress: String,
                       deviceSendProductName: String,
                       deviceSendProductVersion: String,
                       collectorReceiptTime: String,
                       deviceReceiptTime: String,
                       endTime: String,
                       catTechnique: String,
                       catBehavior: String,
                       catOutcome: String,
                       direction: String,
                       eventId: String)

class DBAuditJob {

  def execute(): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    val properties = new Properties()
    properties.setProperty("bootstrap.servers", "192.168.30.79:19092")
    properties.setProperty("zookeeper.connect", "192.168.30.79:12181")

    val mapper: ObjectMapper = new ObjectMapper()


    val dbStream = env.addSource(new FlinkKafkaConsumer[String]("com.dbapp.dbaudit",
      new SimpleStringSchema(), properties))

    //   dbStream.print()
    dbStream.map(mapper.readValue(_ , classOf[DBAuditLog])).print()

    env.execute("Socket Window WordCount")
  }
}

File: stream-db-analyse/src/main/java/com/dbapp/analyse/stream/job/DBAuditJob.java
package com.dbapp.analyse.stream.job;

import com.dbapp.analyse.stream.dto.AdultLogBody;
import com.dbapp.analyse.stream.utils.JacksonJsonUtils;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer;

import java.util.Properties;

public class DBAuditJob {
    public void execute(String jobName) throws Exception {
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        Properties properties = new Properties();
        properties.setProperty("bootstrap.servers", "192.168.30.79:19092");
        properties.setProperty("zookeeper.connect", "192.168.30.79:12181");
        // TODO 记录offset以便下次继续消费

        // TODO check point

        // TODO save point


        DataStream<String> stream = env
                .addSource(new FlinkKafkaConsumer<>("com.dbapp.dbaudit", new SimpleStringSchema(), properties));


       stream.map(json -> JacksonJsonUtils.fromJson(json, AdultLogBody.class)).filter(adultlog->{

           // TODO 过滤掉有缺损的数据
           // TODO 验证server 端口等信息

           return true;
       }).map(e->{
           System.out.println("----------------------------------");
           // TODO 解析结果集
           System.out.println(e.getDataSet());
           return e;
       });


        // get dataSet


        FlinkKafkaProducer<String> myProducer = new FlinkKafkaProducer<>("com.dbapp.dbaudit.out",
                new SimpleStringSchema(), properties);

        // TODO 目前这个是到kafka
        //  TODO 可以手动定义sink保存到图库
        stream.addSink(myProducer);
        env.execute(jobName);
    }
}
